### PR TITLE
[WIP] - Fix multiple extruder issues

### PR DESCRIPTION
**Pull request by diegopradogesto**
Adding some fixes to some problems that exist when switching number of extruders in multiextruder printers. The idea is to keep `active_extruder_index = -1` when there is just one extruder, and then the global stack is taken into account. Also, the `active_extruder_stacks` don't exist in the case we just have one extruder, so the global stack will be used to take the materials, qualities and so on.

There were some points in which the UI is setting the active extruder index value, think that it's consider to be done in python. So all these references were deleted and now it's done through signals.

Those fixes try to solve CURA-4386 and CURA-4379

### CODE REVIEW

File: resources/qml/Menus/MaterialMenu.qml [hunk 1]
```diff
@@ -70,13 +66,7 @@ Menu
             checkable: true
             checked: model.id == Cura.MachineManager.allActiveMaterialIds[ExtruderManager.extruderIds[extruderIndex]]
             exclusiveGroup: group
-            onTriggered:
-            {
-                var activeExtruderIndex = ExtruderManager.activeExtruderIndex;
-                ExtruderManager.setActiveExtruderIndex(extruderIndex);
-                Cura.MachineManager.setActiveMaterial(model.id);
-                ExtruderManager.setActiveExtruderIndex(activeExtruderIndex);
-            }
```
Comment (fieldOfView) [inline resources/qml/Menus/MaterialMenu.qml]: There was a reason this code was in place. See https://github.com/Ultimaker/Cura/pull/2224. It looks like this commit breaks that again.

Comment (diegopradogesto) [reply resources/qml/Menus/MaterialMenu.qml]: Oh, I will take a look at it. Thank you for the comment

Comment (diegopradogesto) [reply resources/qml/Menus/MaterialMenu.qml]: I have reverted your workaround and add some comments. The other issues seem to work as well.

File: cura/Settings/ExtruderManager.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ class ExtruderManager(QObject):
     def __init__(self, parent = None):
         super().__init__(parent)
         self._extruder_trains = { } #Per machine, a dictionary of extruder container stack IDs. Only for separately defined extruders.
-        self._active_extruder_index = 0
+        self._active_extruder_index = -1
```
Comment (nallath) [inline cura/Settings/ExtruderManager.py]: Some documentation would be nice; I assume that we use -1 as "not set" right?

File: cura/Settings/ExtruderManager.py [hunk 3]
```diff
@@ -78,6 +78,7 @@ def extruderCount(self):
     def extruderIds(self):
         map = {}
```
Comment (nallath) [inline cura/Settings/ExtruderManager.py]: Just noticed this; But map shadows a built in. Could this be re-factored to a better name?

File: cura/Settings/ExtruderManager.py [hunk 4]
```diff
@@ -513,17 +514,32 @@ def getActiveExtruderStacks(self) -> List["ExtruderStack"]:
         global_stack = Application.getInstance().getGlobalContainerStack()
 
         result = []
+        machine_extruder_count = global_stack.getProperty("machine_extruder_count", "value")
+
+        if machine_extruder_count is 1:
```
Comment (nallath) [inline cura/Settings/ExtruderManager.py]: is shouldn't be used in this case; should be == 
Also; document why we are aborting already.

File: cura/Settings/ExtruderManager.py [hunk 5]
```diff
@@ -513,17 +514,32 @@ def getActiveExtruderStacks(self) -> List["ExtruderStack"]:
...
             self._global_container_stack_definition_id = global_container_stack.getBottom().getId()
             self.globalContainerStackDefinitionChanged.emit()
+
+        # If the global container changed, the number of extruders could be changed and so the active_extruder_index is updated
+        extruder_count = global_container_stack.getProperty("machine_extruder_count", "value")
+        if extruder_count > 1:
+            if self._active_extruder_index == -1:
+                self.setActiveExtruderIndex(0)
+        else:
+            if self._active_extruder_index > -1:
```
Comment (nallath) [inline cura/Settings/ExtruderManager.py]: Might be a bit clearer if we explictly check for != -1 (As we stated that -1 is a special case)  (minor issue)

**Comment by diegopradogesto** [comment]
Just added some minor changes and commented the code.

**Pull request merged**